### PR TITLE
[Formatter] Add several new node implementations

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose
Adds the following node implementations:

* ImportDeclarationNode
* ServiceDeclarationNode
* ExplicitNewExpressionNode
* ParenthesizedArgList
* ServiceBodyNode
* QualifiedNameReferenceNode
* ReturnTypeDescriptorNode
* OptionalTypeDescriptorNode
* ExpressionStatementNode
* CheckExpressionNode
* RemoteMethodCallActionNode
* SimpleNameReferenceNode
* TypeDefinitionNode
* SingletonTypeDescriptorNode
* WhileStatementNode
* BracedExpressionNode
* AssignmentStatementNode
* CompoundAssignmentStatementNode
* DoStatementNode
* ForEachStatementNode
* BinaryExpressionNode
* OnFailClauseNode
* ReturnStatementNode
* FunctionCallExpressionNode

Partially Fixes #26052

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by SupunS — CHANGES_REQUESTED** [review]

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 1]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+    @Override
+    public ImportDeclarationNode transform(ImportDeclarationNode importDeclarationNode) {
+        Token importKeyword = formatToken(importDeclarationNode.importKeyword(), 0, 0);
+
+        if (importDeclarationNode.orgName().isPresent()) {
+            ImportOrgNameNode orgName = formatNode(importDeclarationNode.orgName().get(), 0, 0);
+            importDeclarationNode = importDeclarationNode.modify().withOrgName(orgName).apply();
+        }
+        SeparatedNodeList<IdentifierToken> moduleNames = formatSeparatedNodeList(importDeclarationNode.moduleName(),
+                0, 0, 0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed.
It is needed if a prefix or version is specified, else it should not be there.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 2]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+
+        if (importDeclarationNode.version().isPresent()) {
+            ImportVersionNode version = formatNode(importDeclarationNode.version().get(), 0, 0);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed to have a trailing WS if there is a prefix defined.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 3]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+            importDeclarationNode = importDeclarationNode.modify().withVersion(version).apply();
+        }
+        if (importDeclarationNode.prefix().isPresent()) {
+            ImportPrefixNode prefix = formatNode(importDeclarationNode.prefix().get(), 0, 0);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:452]: Here a trailing WS is not required AFAIU.

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 4]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+            importDeclarationNode = importDeclarationNode.modify().withPrefix(prefix).apply();
+        }
+        Token semicolon = formatToken(importDeclarationNode.semicolon(), 0, 1);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 5]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
                 .apply();
     }
 
+    @Override
+    public ImportDeclarationNode transform(ImportDeclarationNode importDeclarationNode) {
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:433]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 6]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .withRhsExpression(rhsExpression)
+                .withSemicolonToken(semicolonToken)
+                .apply();
+    }
+
+    @Override
+    public DoStatementNode transform(DoStatementNode doStatementNode) {
+        Token doKeyword = formatToken(doStatementNode.doKeyword(), 1, 0);
+        BlockStatementNode blockStatement = formatNode(doStatementNode.blockStatement(),
+                this.trailingWS, this.trailingNL);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 7]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+        if (serviceDeclarationNode.metadata().isPresent()) {
+            MetadataNode metadata = formatNode(serviceDeclarationNode.metadata().get(), 1, 0);
+            serviceDeclarationNode = serviceDeclarationNode.modify().withMetadata(metadata).apply();
+        }
+
+        Token serviceKeyword = formatToken(serviceDeclarationNode.serviceKeyword(), 1, 0);
+        IdentifierToken serviceName = formatToken(serviceDeclarationNode.serviceName(), 1, 0);
+        Token onKeyword = formatToken(serviceDeclarationNode.onKeyword(), 1, 0);
+        SeparatedNodeList<ExpressionNode> expressions =
+                formatSeparatedNodeList(serviceDeclarationNode.expressions(), 0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 8]
```diff
@@ -399,6 +429,392 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+        return doStatementNode.modify()
+                .withDoKeyword(doKeyword)
+                .withBlockStatement(blockStatement)
+                .apply();
+    }
+
+    @Override
+    public ForEachStatementNode transform(ForEachStatementNode forEachStatementNode) {
+        Token forEachKeyword = formatToken(forEachStatementNode.forEachKeyword(), 1, 0);
+        TypedBindingPatternNode typedBindingPattern = formatNode(forEachStatementNode.typedBindingPattern(), 0, 0);
```
[the remark this answers is not in the document]
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Fixed

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 9]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+    @Override
+    public ForEachStatementNode transform(ForEachStatementNode forEachStatementNode) {
+        boolean hasOnFailClause = forEachStatementNode.onFailClause().isPresent();
+
+        Token forEachKeyword = formatToken(forEachStatementNode.forEachKeyword(), 1, 0);
+        TypedBindingPatternNode typedBindingPattern = formatNode(forEachStatementNode.typedBindingPattern(), 1, 0);
+        Token inKeyword = formatToken(forEachStatementNode.inKeyword(), 1, 0);
+        Node actionOrExpressionNode = formatNode(forEachStatementNode.actionOrExpressionNode(), 1, 0);
+        StatementNode blockStatement = formatNode(forEachStatementNode.blockStatement(),
+                hasOnFailClause ? 1 : this.trailingWS, hasOnFailClause ? 0 : this.trailingNL);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:781]: moving this inside the below `if (hasOnFailClause) {` block would be cleaner..

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 10]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .withColon(colon)
+                .withIdentifier(identifier)
+                .apply();
+    }
+
+    @Override
+    public ReturnTypeDescriptorNode transform(ReturnTypeDescriptorNode returnTypeDescriptorNode) {
+        Token returnsKeyword = formatToken(returnTypeDescriptorNode.returnsKeyword(), 1, 0);
+        NodeList<AnnotationNode> annotations = formatNodeList(returnTypeDescriptorNode.annotations(), 0, 0, 1, 0);
+        Node type = formatNode(returnTypeDescriptorNode.type(), 1, 0);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:581]: should use global var for WS

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 11]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .withExpressions(expressions)
+                .withServiceBody(serviceBody)
+                .apply();
+    }
+
+    @Override
+    public ExplicitNewExpressionNode transform(ExplicitNewExpressionNode explicitNewExpressionNode) {
+        Token newKeywordToken = formatToken(explicitNewExpressionNode.newKeyword(), 1, 0);
+        TypeDescriptorNode typeDescriptorNode = formatNode(explicitNewExpressionNode.typeDescriptor(), 0, 0);
+        ParenthesizedArgList parenthesizedArgList = formatNode(explicitNewExpressionNode.parenthesizedArgList(), 0, 0);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:525]: should use global var for WS

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 12]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+        return importOrgNameNode.modify()
+                .withOrgName(orgName)
+                .withSlashToken(slashToken)
+                .apply();
+    }
+
+    @Override
+    public ImportPrefixNode transform(ImportPrefixNode importPrefixNode) {
+        Token asKeyword = formatToken(importPrefixNode.asKeyword(), 1, 0);
+        Token prefix = formatToken(importPrefixNode.prefix(), 0, 0);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:478]: should use global var for WS

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 13]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .withAsKeyword(asKeyword)
+                .withPrefix(prefix)
+                .apply();
+    }
+
+    @Override
+    public ImportVersionNode transform(ImportVersionNode importVersionNode) {
+        Token versionKeyword = formatToken(importVersionNode.versionKeyword(), 1, 0);
+        SeparatedNodeList<Token> versionNumber = formatSeparatedNodeList(importVersionNode.versionNumber(),
+                0, 0, 0, 0, this.trailingWS, 0);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:490]: should use global var for trailingNL

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 14]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .apply();
+    }
+
+    @Override
+    public OnFailClauseNode transform(OnFailClauseNode onFailClauseNode) {
+        Token onKeyword = formatToken(onFailClauseNode.onKeyword(), 1, 0);
+        Token failKeyword = formatToken(onFailClauseNode.failKeyword(), 1, 0);
+        TypeDescriptorNode typeDescriptor = formatNode(onFailClauseNode.typeDescriptor(), 1, 0);
+        IdentifierToken failErrorName = formatToken(onFailClauseNode.failErrorName(), 1, 0);
+        BlockStatementNode blockStatement = formatNode(onFailClauseNode.blockStatement(), 0, 1);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:817]: should use global var for WS and NL

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java [hunk 15]
```diff
@@ -399,6 +429,436 @@ public RecordFieldWithDefaultValueNode transform(RecordFieldWithDefaultValueNode
...
+                .apply();
+    }
+
+    @Override
+    public DoStatementNode transform(DoStatementNode doStatementNode) {
+        boolean hasOnFailClause = doStatementNode.onFailClause().isPresent();
+
+        Token doKeyword = formatToken(doStatementNode.doKeyword(), 1, 0);
+        BlockStatementNode blockStatement = formatNode(doStatementNode.blockStatement(),
+                hasOnFailClause ? 1 : this.trailingWS, hasOnFailClause ? 0 : this.trailingNL);
```
Comment (SupunS) [inline misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java:758]: moving this inside the below if (hasOnFailClause) { block would be cleaner..

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 9 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 10 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 11 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 12 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 13 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 15 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

File: misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java (hunk 14 above)
Comment (IrushiL) [reply misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/NewFormattingTreeModifier.java]: Done

**Review by SupunS — APPROVED** [review]

**Pull request merged**